OsmAnd/src/net/osmand/plus/plugins/externalsensors/adapters/DevicesForWidgetAdapter.java

```diff
@@ -30,43 +41,68 @@ public void setDeviceId(@Nullable String selectedDeviceId) {
 
 	@Override
 	public void onBindViewHolder(@NonNull FoundDeviceViewHolder holder, int position) {
-		if (position == 0) {
-			holder.selectionMark.setChecked(Algorithms.isEmpty(selectedDeviceId) || ExternalSensorsPlugin.DENY_WRITE_SENSOR_DATA_TO_TRACK_KEY.equals(selectedDeviceId));
+		int itemType = getItemViewType(position);
+		if (itemType == NONE_ITEM_TYPE) {
+			holder.selectionMark.setChecked(Algorithms.isEmpty(selectedDeviceId) || !ANY_CONNECTED_DEVICE_WRITE_SENSOR_DATA_TO_TRACK_KEY.equals(selectedDeviceId) && plugin.getDevice(selectedDeviceId) == null);
 			holder.itemView.setOnClickListener(v -> onItemClicked(holder, null));
 			holder.description.setVisibility(View.GONE);
 			holder.name.setText(R.string.shared_string_none);
-			holder.icon.setImageDrawable(uiUtils.getIcon(R.drawable.ic_action_sensor_off, nightMode));
+			holder.icon.setImageDrawable(uiUtils.getIcon(R.drawable.ic_action_track_disabled, nightMode));
+		} else if (itemType == ANY_CONNECTED_ITEM_TYPE) {
+			holder.selectionMark.setChecked(ANY_CONNECTED_DEVICE_WRITE_SENSOR_DATA_TO_TRACK_KEY.equals(selectedDeviceId));
+			holder.itemView.setOnClickListener(v -> onItemClicked(holder, ANY_CONNECTED_DEVICE_WRITE_SENSOR_DATA_TO_TRACK_KEY));
+			holder.description.setVisibility(View.GONE);
+			holder.name.setText(R.string.any_connected);
+			holder.icon.setImageDrawable(uiUtils.getIcon(widgetDataFieldType.getIconId(), nightMode));
 		} else {
-			super.onBindViewHolder(holder, position - 1);
+			super.onBindViewHolder(holder, position - getNonDeviceItemsCount());
 			holder.description.setVisibility(View.VISIBLE);
 			holder.menuIcon.setVisibility(View.GONE);
-			AbstractDevice<?> device = items.get(position - 1);
+			AbstractDevice<?> device = items.get(position - getNonDeviceItemsCount());
 			holder.selectionMark.setChecked(Algorithms.stringsEqual(selectedDeviceId, device.getDeviceId()));
-			holder.itemView.setOnClickListener(v -> onItemClicked(holder, device));
+			holder.itemView.setOnClickListener(v -> onItemClicked(holder, device.getDeviceId()));
 		}
 		if (holder.selectionMark.isChecked()) {
 			setSelectedPosition(position);
 		}
 		holder.selectionMark.setVisibility(View.VISIBLE);
 	}
 
+	@Override
+	public int getItemViewType(int position) {
+		if (position == 0) {
+			return withNoneVariant ? NONE_ITEM_TYPE : ANY_CONNECTED_ITEM_TYPE;
+		} else if (position == 1) {
+			return withNoneVariant ? ANY_CONNECTED_ITEM_TYPE : DEVICE_ITEM_TYPE;
```

Comment:
Looks complicated, probably better to add constants to items in the adapter, similar to [ImportTracksAdapter](https://github.com/osmandapp/Osmand/blob/a174c86ee49cca0bb6af2895cd2bb0afda0468dc/OsmAnd/src/net/osmand/plus/importfiles/ui/ImportTracksAdapter.java#L128)